packages/protobuf/src/extension-accessor.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright 2021-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { Message } from "./message.js";
+import type { BinaryReadOptions, BinaryWriteOptions } from "./binary-format.js";
+import { assert } from "./private/assert.js";
+import {
+  createExtensionContainer,
+  filterUnknownFields,
+} from "./private/extensions.js";
+import type { Extension } from "./extension.js";
+
+/**
+ * Retrieve an extension value from a message.
+ *
+ * The function never returns undefined. Use hasExtension() to check whether an
+ * extension is set. If the extension is not set, this function returns the
+ * default value (if one was specified in the protobuf source), or the zero value
+ * (for example `0` for numeric types, `[]` for repeated extension fields, and
+ * an empty message instance for message fields).
+ */
+export function getExtension<E extends Message<E>, V>(
```

Comment:
A consequential downside of this approach that may need considerable documentation is that extension values are effectively read-only. So even if this returns a mutable value, mutating it has no effect on the original extended message.  Instead, `setExtension` must always be called, even if the top-level element (like an array or object) is unchanged but its contents are being mutated. This may be counter-intuitive compared to all other kinds of field access on the message.

Reply:
👍 Added a comment in https://github.com/bufbuild/protobuf-es/pull/666/commits/5fd6e9635560735be3a8c8b0c783130f3ab6fa71.

I think the documentation (a follow-up to this PR) will make this clear via examples, and I'll make sure to call it out explicitly.